### PR TITLE
Actor Sheet - Safely Handle user generated JS

### PR DESCRIPTION
**Pull request by laura-wenning**
Changes
- Adds a Web Worker for safely running user-generated JS within string literals
- Changes how variables are accessed within the actor sheet from `{$variableName}` to `${variableName}` to be immediately runnable using eval()

Notes
- This iteration is pretty slow and needs a fix for that in the near future

Depends on #253

### CODE REVIEW

File: src/@owl-factory/web-worker/components/ExampleWorker.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import React from "react";
+import { newWebWorker } from "../utilities";
+import ExampleSandboxWorker from "../workers/sandbox-example.worker";
+
+export function ExampleWorker() {
```
Comment (laura-wenning) [inline src/@owl-factory/web-worker/components/ExampleWorker.tsx:5]: Comment why this is here

File: src/@owl-factory/web-worker/types/index.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export type OnMessage = (
+  ((this: WindowEventHandlers, ev: MessageEvent<any>) => any) &
+  ((this: Window, ev: MessageEvent<any>) => any)
+) | null;
+
+export interface SandboxWorkerMethods {
```
Comment (laura-wenning) [inline src/@owl-factory/web-worker/types/index.ts:6]: Is this used anywhere?

File: src/@owl-factory/web-worker/workers/example.worker.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Calculates the nth number in the fibbonacci sequence
```
Comment (laura-wenning) [inline src/@owl-factory/web-worker/workers/example.worker.ts:2]: Comment better to describe what this is doing

File: src/@owl-factory/web-worker/components/ExampleWorker.tsx (hunk 1 above)
Comment (laura-wenning) [inline src/@owl-factory/web-worker/components/ExampleWorker.tsx:5]: Comment better for what this does

File: src/nodes/actor-sheets/controllers/ActorController.ts [hunk 4]
```diff
@@ -320,24 +335,36 @@ class $ActorController {
...
+    expr: ParsedExpressionString,
+    properties: SheetProperties
+  ): string {
+    if (!expr.isExpression) { return expr.value; }
+    const key = `${elementKey}__${fieldName}`;
+    // console.log(key)
+    const { actorRef, sheetRef, rulesetRef } = this.$renders[renderID];
+    const message: SandboxWorkerMessage = {
+      ...properties,
+      actor: toJS(this.actorController.getActorValues(actorRef)),
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/controllers/ActorController.ts:355]: This is really slow but we need a new PR to fix this

File: src/nodes/actor-sheets/controllers/ActorController.ts [hunk 5]
```diff
@@ -358,6 +385,12 @@ class $ActorController {
     return renderedResult;
   }
 
+  public setVariables(msg: any) {
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/controllers/ActorController.ts:388]: Comment

File: src/nodes/actor-sheets/controllers/ActorController.ts [hunk 6]
```diff
@@ -410,3 +443,8 @@ class $ActorController {
 }
 
 export const ActorController = new $ActorController();
+
+function getFieldKey<T extends GenericSheetElementDescriptor>(element: T, fieldName: string) {
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/controllers/ActorController.ts:447]: Comment

File: src/nodes/actor-sheets/controllers/ActorSubController.ts [hunk 7]
```diff
@@ -80,6 +80,11 @@ export class ActorSubController {
     this.$actorValues[ref][fieldName] = value;
   }
 
+  public getAllContent(ref: string): Record<string, ActorContent[]> {
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/controllers/ActorSubController.ts:83]: Comment

File: src/nodes/actor-sheets/controllers/SheetController.ts [hunk 8]
```diff
@@ -166,6 +170,12 @@ export class SheetController<T> {
     return this.tabs[sheetID][key];
   }
 
+  public getAllVariables(key: string) {
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/controllers/SheetController.ts:173]: Comment

File: src/nodes/actor-sheets/types/workers.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { Scalar } from "types";
+import { ActorContent } from "types/documents/Actor";
+import { StaticVariable } from "types/documents/subdocument/StaticVariable";
+
+export type SandboxWorkerMessage = StaticSandboxWorkerMessage & Record<string, Scalar | unknown>;
+
+export interface StaticSandboxWorkerMessage {
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/types/workers.ts:7]: Comment

File: src/nodes/actor-sheets/utilities/expressions/parse.ts [hunk 10]
```diff
@@ -11,34 +11,40 @@ const EXPRESSION_PART_REGEXES = [
...
-    i++; if (i>1000) { break; }
-  }
-
-  return exprs;
+  const value: ParsedExpressionString = {
+    value: str,
+    isExpression: str.search(/^\$\{|[^\\]\$\{/) !== -1,
+  };
+  return value;
+
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/utilities/expressions/parse.ts:19]: Remove

File: src/nodes/actor-sheets/workers/sandboxed-code.worker.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    "console": 1,
+  };
+
+  // Loops through all of the property names for the global object. For anything that is not whitelisted,
+  //  redefine to throw an error instead
+  Object.getOwnPropertyNames( global ).forEach( function(prop) {
+    // If the whitelist has this property, do nothing
+    if (Object.prototype.hasOwnProperty.call(wl, prop)) { return; }
+
+    // Object.defineProperty(global, prop, {
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/workers/sandboxed-code.worker.ts:55]: Uncomment this

File: src/nodes/actor-sheets/workers/sandboxed-code.worker.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    // });
+  });
+
+  // Loops through all of the property names for the global object. For anything that is not whitelisted,
+  //  redefine to throw an error instead
+  Object.getOwnPropertyNames( (global as any).__proto__ ).forEach( function(prop) {
+    // If the whitelist has this property, do nothing
+    if (Object.prototype.hasOwnProperty.call(wl, prop)) { return; }
+
+    // Object.defineProperty((global as any).__proto__, prop, {
```
Comment (laura-wenning) [inline src/nodes/actor-sheets/workers/sandboxed-code.worker.ts:69]: Uncomment this

File: src/@owl-factory/web-worker/types/index.ts (hunk 2 above)
Comment (laura-wenning) [reply src/@owl-factory/web-worker/types/index.ts:6]: It is not. Deleted

File: src/nodes/actor-sheets/types/workers.ts (hunk 9 above)
Comment (laura-wenning) [reply src/nodes/actor-sheets/types/workers.ts:7]: Deleted due to unuse

Comment (laura-wenning) [reply src/nodes/actor-sheets/types/workers.ts:7]: I lied. It is used. Restored and commented

**Review by lucyawrey — APPROVED** [review]

**Pull request merged**